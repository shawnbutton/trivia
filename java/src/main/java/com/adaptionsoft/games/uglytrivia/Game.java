package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Game {
    public static final int NUMBER_OF_QUESTIONS = 50;

    List players = new ArrayList();
    int[] places = new int[6];
    int[] purses = new int[6];
    boolean[] inPenaltyBox = new boolean[6];

    LinkedList popQuestions = new LinkedList();
    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();

    int currentPlayer = 0;
    boolean isGettingOutOfPenaltyBox;

    public Game() {
        for (int i = 0; i < NUMBER_OF_QUESTIONS; i++) {
            logMessage("Adding game question");
            popQuestions.addLast("Pop Question " + i);
            scienceQuestions.addLast(("Science Question " + i));
            sportsQuestions.addLast(("Sports Question " + i));
            rockQuestions.addLast("Rock Question " + i);
        }
    }

    public boolean isPlayable() {
        return (howManyPlayers() >= 2);
    }

    public boolean add(String playerName) {


        players.add(playerName);
        initPlayerStats();

        logMessage(playerName + " was added");
        logMessage("They are player number " + players.size());
        return true;
    }

    private void initPlayerStats() {
        places[howManyPlayers()] = 0;
        purses[howManyPlayers()] = 0;
        inPenaltyBox[howManyPlayers()] = false;
    }

    public int howManyPlayers() {
        return players.size();
    }

    public void roll(int roll) {
        logMessageForCurrentPlayer(" is the current player");
        logMessage("They have rolled a " + roll);

        if (currentPlayerInPenaltyBox()) {
            if (notEven(roll)) {
                isGettingOutOfPenaltyBox = true;

                logMessageForCurrentPlayer(" is getting out of the penalty box");
                incrementCurrentPlayerPlace(roll);

                logMessageForCurrentPlayer("'s new location is " + places[currentPlayer]);
                logMessage("The category is " + currentCategory());
                askQuestion();
            } else {
                logMessageForCurrentPlayer(" is not getting out of the penalty box");
                isGettingOutOfPenaltyBox = false;
            }

        } else {
            incrementCurrentPlayerPlace(roll);

            String message = "'s new location is " + places[currentPlayer];
            logMessageForCurrentPlayer(message);
            logMessage("The category is " + currentCategory());
            askQuestion();
        }

    }

    private boolean currentPlayerInPenaltyBox() {
        return inPenaltyBox[currentPlayer];
    }

    private void logMessageForCurrentPlayer(String message) {
        logMessage(messageForCurrentPlayer(message));
    }

    private boolean notEven(int roll) {
        return roll % 2 != 0;
    }

    private String messageForCurrentPlayer(String message) {
        return currentPlayer() + message;
    }

    private void logMessage(String x) {
        System.out.println(x);
    }

    private void incrementCurrentPlayerPlace(int roll) {
        places[currentPlayer] = places[currentPlayer] + roll;
        if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;
    }

    private Object currentPlayer() {
        return players.get(currentPlayer);
    }

    private void askQuestion() {
        if (currentCategory() == "Pop")
            System.out.println(popQuestions.removeFirst());
        if (currentCategory() == "Science")
            System.out.println(scienceQuestions.removeFirst());
        if (currentCategory() == "Sports")
            System.out.println(sportsQuestions.removeFirst());
        if (currentCategory() == "Rock")
            System.out.println(rockQuestions.removeFirst());
    }


    private String currentCategory() {
        if (places[currentPlayer] == 0) return "Pop";
        if (places[currentPlayer] == 4) return "Pop";
        if (places[currentPlayer] == 8) return "Pop";
        if (places[currentPlayer] == 1) return "Science";
        if (places[currentPlayer] == 5) return "Science";
        if (places[currentPlayer] == 9) return "Science";
        if (places[currentPlayer] == 2) return "Sports";
        if (places[currentPlayer] == 6) return "Sports";
        if (places[currentPlayer] == 10) return "Sports";
        return "Rock";
    }

    public boolean wasCorrectlyAnswered() {
        if (currentPlayerInPenaltyBox()) {
            if (isGettingOutOfPenaltyBox) {
                logMessage("Answer was correct!!!!");
                purses[currentPlayer]++;
                logMessageForCurrentPlayer(" now has " + purses[currentPlayer] + " Gold Coins.");

                boolean winner = didPlayerWin();
                incrementPlayer();

                return winner;
            } else {
                incrementPlayer();
                return true;
            }


        } else {

            logMessage("Answer was corrent!!!!");
            purses[currentPlayer]++;
            logMessageForCurrentPlayer(" now has " + purses[currentPlayer] + " Gold Coins.");

            boolean winner = didPlayerWin();
            incrementPlayer();

            return winner;
        }
    }

    private void incrementPlayer() {
        currentPlayer++;
        if (currentPlayer == players.size()) currentPlayer = 0;
    }

    public boolean wrongAnswer() {
        logMessage("Question was incorrectly answered");
        logMessageForCurrentPlayer(" was sent to the penalty box");
        inPenaltyBox[currentPlayer] = true;

        incrementPlayer();
        return true;
    }


    private boolean didPlayerWin() {
        return !(purses[currentPlayer] == 6);
    }
}
