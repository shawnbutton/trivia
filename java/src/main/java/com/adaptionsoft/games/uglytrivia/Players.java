package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.List;

public class Players {

    List<Player> players = new ArrayList<Player>();

    int currentPlayer = 0;

    public void add(Player player) {
        players.add(player);

        logMessage(player.getName() + " was added");
        logMessage("They are player number " + players.size());

    }

    private void logMessage(String x) {
        System.out.println(x);
    }


    public int size() {
        return players.size();
    }

    public Player get(int currentPlayer) {
        return players.get(currentPlayer);
    }

    public void incrementPlayer() {
        currentPlayer++;
        if (currentPlayer == players.size()) currentPlayer = 0;
    }


    public Player getCurrentPlayer() {
        return players.get(currentPlayer);
    }
}
