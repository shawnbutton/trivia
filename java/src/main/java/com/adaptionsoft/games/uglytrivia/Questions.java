package com.adaptionsoft.games.uglytrivia;

import java.util.LinkedList;

public class Questions {

    LinkedList popQuestions = new LinkedList();
    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();


    public Questions() {
        loadQuestions();
    }

    private void loadQuestions() {
        for (int i = 0; i < 50; i++) {
            logMessage("Adding game question");
            popQuestions.addLast("Pop Question " + i);
            scienceQuestions.addLast(("Science Question " + i));
            sportsQuestions.addLast(("Sports Question " + i));
            rockQuestions.addLast("Rock Question " + i);
        }
    }

    public String getCurrentCategory(int place) {
        if ((place == 0) || (place == 4) || (place == 8)) return "Pop";
        if ((place == 1) || (place == 5)  || (place == 9)) return "Science";
        if ((place == 2)  || (place == 6)  || (place == 10)) return "Sports";
        return "Rock";
    }


    public void askQuestions(String currentCategory) {
        if (currentCategory == "Pop")
            System.out.println(popQuestions.removeFirst());
        if (currentCategory == "Science")
            System.out.println(scienceQuestions.removeFirst());
        if (currentCategory == "Sports")
            System.out.println(sportsQuestions.removeFirst());
        if (currentCategory == "Rock")
            System.out.println(rockQuestions.removeFirst());
    }


    private void logMessage(String x) {
        System.out.println(x);
    }



}
