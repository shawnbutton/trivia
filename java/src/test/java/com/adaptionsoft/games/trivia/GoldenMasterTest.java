package com.adaptionsoft.games.trivia;

import com.adaptionsoft.games.uglytrivia.Game;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class GoldenMasterTest {

    @Test
    public void test_can_set_up_game() {

        Game game = new Game();

        game.add("player 1");
        assertFalse(game.isPlayable());

        game.add("player 2");
        assertTrue(game.isPlayable());

        game.add("player 3");
        assertThat(game.howManyPlayers(), is(3));

    }

    @Test
    public void large_run_should_match_golden_master() throws Exception {


        OutputStream outputStream = redirectSystemOutToStream();

        Game game = new Game();

        game.add("player 1");
        game.add("player 2");
        game.add("player 3");

        game.roll(1);

        System.out.flush();

        Approvals.verify(outputStream.toString());

    }

    private OutputStream redirectSystemOutToStream() {
        OutputStream outputStream = new ByteArrayOutputStream();
        PrintStream ps = new PrintStream(outputStream);
        System.setOut(ps);
        return outputStream;
    }


}
