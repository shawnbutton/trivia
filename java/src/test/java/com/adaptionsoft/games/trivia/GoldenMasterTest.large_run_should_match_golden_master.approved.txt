player 1 was added
They are player number 1
player 2 was added
They are player number 2
player 3 was added
They are player number 3
player 1 is the current player
They have rolled a 1
player 1's new location is 1
The category is Science
Science Question 0
